### PR TITLE
cmake: Toolchain abstraction: Linker abstraction Part 1

### PR DESCRIPTION
**Pull request by mped-oticon**
This is Part 1 of Linker abstraction.
A few commits to start on the Linker part of Toolchain abstraction.
(The last few compiler parts are being worked on).

No functional change expected.

This is motivated by the wish to abstract Zephyr's usage of toolchains,
permitting non-intrusive porting to other (commercial) toolchains.

Zephyr's cmake build system currently assumes GNU ld is used as the linker.
This is not the case for Oticon which uses another incompatible linker, and our own custom linker scripts.
Without these PRs, Oticon and others will have to keep patching Zephyr Cmake code.

Signed-off-by: Mark Ruvald Pedersen <mped@oticon.com>

-------------

Part of #16031

### CODE REVIEW

**Review by SebastianBoe — APPROVED** [review]

**Review by nashif — APPROVED** [review]

File: cmake/target_toolchain.cmake [hunk 1]
```diff
@@ -42,20 +42,20 @@ endif()
 unset(CMAKE_C_COMPILER)
 unset(CMAKE_C_COMPILER CACHE)
 
-# Configure the toolchain based on what toolchain technology is used
-# (gcc, host-gcc etc.)
+# A toolchain consist of a compiler and a linker.
```
Comment (marc-hb) [inline cmake/target_toolchain.cmake]: I think that's a bit too much of a simplification. Looking for instance at the Zephyr SDK: - as, - cpp, - strip, - nm, - objdump, - ar, - ranlib, - gdb, - libgcc.a,  - crtfastmath.o, - crt*.o,
Suggest: "the two main components of a toolchain are the compiler and the linker."

Maybe sneak that rephrase in a Part N not merged yet?

PS: I'm usually the one complaining about how bundling unrelated changes in the same commit breaks cherry-pick etc., however separate commits for small whitespace/comment fixes is maybe it a bit too far?

File: cmake/target_toolchain.cmake [hunk 2]
```diff
@@ -42,20 +42,20 @@ endif()
...
 # Uniquely identify the toolchain wrt. it's capabilities.
 #
 # What we are looking for, is a signature definition that is defined
 # like this:
-
-# Toolchains with the same signature will always support the same set
-# of flags.
-
+#  * Toolchains with the same signature will always support the same set
+#    of flags.
```
Comment (marc-hb) [inline cmake/target_toolchain.cmake]: Good opportunity to add the linker in the signature/checksum @SebastianBoe ?

**Review by marc-hb — COMMENTED** [review]
(Late to the party)

File: cmake/target_toolchain.cmake (hunk 1 above)
Comment (mped-oticon) [reply cmake/target_toolchain.cmake]: Good comment and you are right. These other tools of the toolchain you mention have also been thought of; the will get their own bintools or binutils category of their own.

File: cmake/target_toolchain.cmake (hunk 2 above)
Comment (SebastianBoe) [reply cmake/target_toolchain.cmake]: AFAICT this is correct, the linker should be a part of the toolchain signature.

**Pull request merged**